src/main/cpp/faultinj/faultinj.cu

```diff
@@ -0,0 +1,515 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <assert.h>
+#include <cuda.h>
+#include <cupti.h>
+#include <iostream>
+#include <map>
+#include <pthread.h>
+#include <exception>
+
+// thread-safe ptree
+#define BOOST_SPIRIT_THREADSAFE
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/property_tree/ptree.hpp>
+
+#include <spdlog/spdlog.h>
+#include <sys/inotify.h>
+#include <sys/time.h>
+
+namespace {
+
+#define CUPTI_CALL(call)                                                       \
+do {                                                                           \
+    CUptiResult _status = call;                                                \
+    if (_status != CUPTI_SUCCESS) {                                            \
+        const char *errstr;                                                    \
+        cuptiGetResultString(_status, &errstr);                                \
+        spdlog::error("function {} failed with error {}", #call, errstr);      \
+    }                                                                          \
+} while (0)
+
+#define PTHREAD_CALL(call)                                                         \
+do {                                                                               \
+    int _status = call;                                                            \
+    if (_status != 0) {                                                            \
+        spdlog::error("function {} failed with error code {}", #call, _status);    \
+    }                                                                              \
+} while (0)
+
+typedef enum {
+    FI_TRAP,
+    FI_ASSERT,
+    FI_RETURN_VALUE
+} FaultInjectionType;
+
+typedef struct {
+    volatile uint32_t initialized;
+    CUpti_SubscriberHandle  subscriber;
+
+    std::string configFilePath = std::string();
+
+    bool dynamic;
+    int terminateThread;
+    pthread_t dynamicThread;
+    // TODO change to the RAII idiom
+    pthread_rwlock_t configLock = PTHREAD_RWLOCK_INITIALIZER;
+
+    boost::property_tree::ptree configRoot;
+    boost::optional<boost::property_tree::ptree&> driverFaultConfigs = boost::none;
+    boost::optional<boost::property_tree::ptree&> runtimeFaultConfigs = boost::none;
+} injGlobalControl;
+injGlobalControl globalControl;
+
+// Function Declarations
+void atExitHandler(void);
+
+void CUPTIAPI faultInjectionCallbackHandler(void *userdata, CUpti_CallbackDomain domain, CUpti_CallbackId cbid, void *cbInfo);
+
+const std::string configFilePathEnv = "FAULT_INJECTOR_CONFIG_PATH";
+CUptiResult cuptiInitialize(void);
+void readFaultInjectorConfig(void);
+void traceConfig(boost::property_tree::ptree const& pTree);
+void *dynamicReconfig(void *args);
+
+void
+globalControlInit(void) {
```

Comment:
Style needs to be consistent. As a more widely used style: please do not break line between return type and function name.

Reply:
I wonder if you use any formatter? We should use clang-format.

Reply:
We need pre-commit hool like cudf to do that like in cudf.

Reply:
I would recommend to use the clang-format config from cudf. Our IDE can just load the config and do its job.

Reply:
any style works for me, especially if it is just done automatically. Tieing it to the editor is ok but not ideal.

Reply:
any style works for me, especially if it is just done automatically. Tieing it to the editor is ok but not ideal.